### DISCUSSION TITLE
`pyvista` not at MAJOR version 1?

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvista#2713] [Ideas]
@pyvista/developers Why is `pyvista` not yet at MAJOR version 1? I might be misunderstanding something, so forgive me if this is a silly question. Are we "supposed" to be at 1.0.0 since we have a released API? Granted, the versioning strategy is entirely up to the author and maintainers/developers, but I thought per [semver](https://semver.org/) that once you have a released API you should be at version 1.0.0.

Does `pyvista` follow a different strategy? How/when are versions bumped?

**Comment by akaszynski** [comment]
Good question.

Probably the biggest reason why we're not at 1.0.0 is because we want the flexibility to make significant changes to our API. According to [semver](https://semver.org/)

> 4. Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

Despite the fact that `pyvista` has been public since 12 May 2019, (3 July 2017 was when `vtki` was created), I still feel like we're still getting our footing in regards to our API and I'd for it to remain flexible moving forward. While we're not as popular as `scipy`, according to [scipy release history](https://pypi.org/project/scipy/#history), they took 11 years to get to `1.0.0`.

That being said, I'm open to the suggestion that we finally release a 1.0.0 release, especially with `scipy` coming up and the fact that we don't have any major API changes on the horizon.

---
I just read a decent article criticizing `semver`, [Semantic Versioning Will Not Save You](https://hynek.me/articles/semver-will-not-save-you/), and one of the criticisms is that a consequence of `semver`'s dictation of backwards incompatible changes requiring a bump in the MAJOR version is packages remain as `ZeroVer`:

> Maintainers often feel like there’s an obligation to do SemVer. Some of that is self-imposed by not knowing about alternatives or thinking that “it’s the way it’s done”. Some of that is the result of demands by entitled users. And while SemVer promises freedom (technically you can break compatibility with every release, as long as you increment your major version!), in reality it delivers additional pressure and work.

> The observable result of that pressure is what’s tongue-in-cheek called [0-based Versioning](https://0ver.org/).

> I’ve mentioned that in SemVer the maintainer can do whatever they want, as long as the major version is zero. That leads to many maintainers sticking to their beloved 0ver forever.

PyVista might have fallen into that category, and I'm open to getting 1.0.0 out. However, that means that changes like `point_arrays` --> `point_data` would have required a bump in the MAJOR version. I think if we're fine with permitting some drastic API changes between minor releases, then a bump to 1.0.0 is fine. If not, we might need a few more years before the API is stable enough.

Other opinions welcome here!

**Reply by adam-grant-hendry** [reply]
Fascinating reads! I definitely learned a lot.

I wasn't aware of
- [0ver](https://0ver.org/)

and (except for [flit](https://flit.pypa.io/en/latest/)), I wasn't aware of these alternative dependency managers

- [pdm](https://pdm.fming.dev/#vscode)

Before reading, I thought `semver` could be used as a reliable contract to end-users to identify what changes had been made and whether users should include them in their work (specifically, with a dependency manager, which was the entire reason `semver` appealed to me: an agreed-upon version string syntax it could parse and the ability to pin versions).

After reading, I realized

1. [Hyrum's Law](https://www.hyrumslaw.com/) indicates you cannot determine a priori what changes will be breaking for the end-user; you can only specify what you as the developer intend should be breaking or non-breaking (end users still have to test)

1. There's an underlying assumption that major versions will be maintained as long-lived releases, perhaps in a separate branch. 
   - Hence, if you anticipate you'll add many new non-backward-compatible additions/removals to your codebase, you will be forced to create and maintain many releases.
   - You can't continue the `0`-branch for development, since `0` is older than `1`, and new versions are required to have a higher number in semver. Paradoxically, you could tell users to "pull from main/master" as the newest branch, but it (consequently) won't have a version number to parse!

In fact, on the latter point, [Brett Cannon](https://snarky.ca/why-i-dont-like-semver/) makes a good argument:

> I had a bit of an epiphany while thinking on this topic: version numbers are just a mapping of a sequence of digits to our branching strategy in source **control.**

This is so true. It feels odd to me (since I've become aesthetically accustomed to an `X.Y.Z` version string), but since we only have a main branch, it seems to me we should have a single number for our version (namely `X`), which [PEP440](https://www.python.org/dev/peps/pep-0440/#public-version-identifiers) supports.

> That being said, I'm open to the suggestion that we finally release a 1.0.0 release, especially with scipy coming up and the fact that we don't have any major API changes on the horizon.

If we are following semver, I think this depends on whether we want to keep the current state of the project as a long-lived maintained branch. And, 

- We'd have to agree on how long we would like to keep the branch maintained
- This may open up the question of how often we want to release new major versions (which honestly seems arbitrary to me now based on my previous reading)

Honestly, I'm not sure there is any need from the end-users yet to keep a maintained long-lived branch.

Also, there are at least 2 other things I would like to add:

- Implement the fix I posted in [Issue #500](https://github.com/pyvista/pyvista-support/issues/500)
- Add [`vmtk`](http://www.vmtk.org/) to the library

**Reply by adam-grant-hendry** [reply]
I also appreciate this quote from [Donald Stufft](https://caremad.io/posts/2016/02/versioning-software/), which I agree is at the heart of versioning:

> At the end of the day, the most important, but often overlooked, aspect of the version is that one of it’s primary purposes is to communicate with the end users of the software. It’s up to the project what information is most important to their end users that it deserves to be communicated through the version number, though I would suggest that for most projects, SemVer-ish is likely to be best suited.

**Reply by akaszynski** [reply]
> ... SemVer-ish is likely to be best suited.

I think this is the ideal version control pattern for pyvista, and I envision us remaining in 0.X.Y a bit longer until we're reasonably sure the API won't undergo anymore drastic changes.

**Reply by banesullivan** [reply]
Drive-by commenting as there is a ton of excellent content above that I don't have time to respond fully to...

> I envision us remaining in 0.X.Y a bit longer until we're reasonably sure the API won't undergo anymore drastic changes.

+1. We're not quite where we need to be to release a v1 yet from my perspective.

On my:

> where we need to be

This is a can of worms 🪱 but in brief, we need to be "stable" and not have tons of open issues proposing API changes. Too much is getting changed too often and I like the flexibility that being sub-v1 gives us

---

There is a future where I can imagine us releasing a v1 as a refactored version of PyVista that uses some interesting new features coming soon to VTK Python, but I'll save that for when it's ready 😉

**Reply by adam-grant-hendry** [reply]
Completely agree. The articles opened my eyes to the fact that one can only claim a change to be breaking or non-breaking with respect to the library API and test suite. It is impossible to guarantee something is non-breaking for the end user, which is the irony of semver because it intends version numbers to be customer facing (i.e. a sort of "promise" or indication to the end user).

Even in the best-maintained repos, a change to a dependency claiming to be non-breaking will eventually find a way to break the repo due to [Hyrum's Law](https://www.hyrumslaw.com/). A robust test suite will help thwart this, but at the end of the day a 3rd party won't always document things fully or how their end-users may use the package (in ways they didn't predict).

I think a PEP 440 semverish approach is the wise move. Interestingly, semver (to my knowledge) doesn't really talk about how MAJOR revs are intended to be long-lived branches. Instead of "breaking"/"non-breaking" changes, I think semver should instead focus on "long maintained" vs "rapid iteration". I would vote that a project should bump to 1 when there are enough features end users want to keep and are stable and there are enough maintainers to support it long-term.
